packages/get-starknet/src/wallet.ts

```diff
@@ -220,13 +236,108 @@ export class MetaMaskSnapWallet implements StarknetWindowObject {
     return true;
   }
 
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  on<Event extends keyof WalletEventHandlers>(_event: Event, _handleEvent: WalletEventHandlers[Event]): void {
-    // No operation for now
+  /**
+   * Registers an event handler for `accountsChanged` or `networkChanged`.
+   *
+   * @param event - The event name ('accountsChanged' or 'networkChanged')
+   * @param handleEvent - The event handler function
+   */
+  on<Event extends keyof WalletEventHandlers>(event: Event, handleEvent: WalletEventHandlers[Event]): void {
+    if (!this.#pollingController) {
+      this.#startPolling();
+    }
+
+    if (event === 'accountsChanged') {
+      this.#accountChangeHandlers.push(handleEvent as AccountChangeEventHandler);
+    } else if (event === 'networkChanged') {
+      this.#networkChangeHandlers.push(handleEvent as NetworkChangeEventHandler);
+    } else {
+      throw new Error(`Unsupported event: ${String(event)}`);
+    }
+  }
+
+  /**
+   * Removes an event handler for `accountsChanged` or `networkChanged`.
+   *
+   * @param event - The event name ('accountsChanged' or 'networkChanged')
+   * @param _handleEvent
+   */
+  off<Event extends keyof WalletEventHandlers>(event: Event, _handleEvent?: WalletEventHandlers[Event]): void {
+    if (event === 'accountsChanged') {
+      this.#accountChangeHandlers = [];
+    } else if (event === 'networkChanged') {
+      this.#networkChangeHandlers = [];
+    } else {
+      throw new Error(`Unsupported event: ${String(event)}`);
+    }
+
+    if (this.#accountChangeHandlers.length === 0 && this.#networkChangeHandlers.length === 0) {
+      this.#stopPolling();
+    }
+  }
+
+  /**
+   * Polling function for detecting changes with a maximum delay.
+   * Ensures that each iteration respects a maximum delay before proceeding.
+   */
+  #pollingFunction = async (): Promise<void> => {
+    if (!this.#pollingController) {
+      return; // Abort if the polling controller is not initialized
+    }
+    const { signal } = this.#pollingController;
+
+    while (!signal.aborted) {
+      const startTime = Date.now(); // Track the start time of the iteration
+      const previousNetwork = this.#network;
+      const previousAddress = this.#latestAddress;
+
+      try {
+        // Perform the polling operation with a timeout
+        await Promise.race([
+          this.#init(), // Perform the main operation
+          new Promise(
+            (_, reject) =>
+              setTimeout(() => reject(new Error('Polling timeout exceeded')), MetaMaskSnapWallet.maximumPollingDelay), // Timeout after 5 seconds
+          ),
+        ]);
+
+        // Check for network change
+        if (previousNetwork.chainId !== this.#network.chainId) {
+          this.#networkChangeHandlers.forEach((callback) => callback(this.#network.chainId, [this.#selectedAddress]));
+        }
```

Comment:
i guess you dont wanna break if any of handler fail
and it also better running in a "async" way ?

assume u have some function like below 
```typescript
const resolver = (func, arg) => {
  return new Promise((resolve) => {  
    resolve(func(arg))
  }); 
}
```
in that case, it can support either async or non-async, wont break and run all function at once
(for better design, please use batch call for max 50 calls in one go, please refer to [here](https://github.com/MetaMask/snap-bitcoin-wallet/blob/main/packages/snap/src/utils/async.ts))



```suggestion
        if (previousNetwork.chainId !== this.#network.chainId) {
          await Promise.allSettled(this.#networkChangeHandlers.map(
               (callback) => resolver(callback, this.#selectedAddress)
          ))
        }
```